### PR TITLE
fix: Enable reconnectingpty loadtest and fix/improve logging

### PR DESCRIPTION
**Pull request by mafredri**
This commit re-enabled reconnectingpty loadtests after a logging
refactor of `(*agent).handleReconnectingPTY`. The reasons the tests were
flaking was that `logger.Error` was being called and `slogtest` failing
the test.

We could have set the option for `slogtest` to disable failing, but that
could hide real issues. The current approach improves reconnectingpty
logging overall and provides more insight into what's happening. It's
expected that reconnectingpty sessions fail after the agent is closed,
so calling `logger.Error` at that point is not wanted.

Ref: #5322

### CODE REVIEW

File: agent/agent.go [hunk 1]
```diff
@@ -889,23 +912,22 @@ func (a *agent) handleReconnectingPTY(ctx context.Context, msg codersdk.Reconnec
...
 		// We can continue after this, it's not fatal!
-		a.logger.Error(ctx, "resize reconnecting pty", slog.F("id", msg.ID), slog.Error(err))
+		logger.Error(ctx, "resize", slog.Error(err))
 	}
 	// Write any previously stored data for the TTY.
 	rpty.circularBufferMutex.RLock()
-	_, err = conn.Write(rpty.circularBuffer.Bytes())
+	prevBuf := slices.Clone(rpty.circularBuffer.Bytes())
 	rpty.circularBufferMutex.RUnlock()
+	_, err = conn.Write(prevBuf)
```
Comment (mafredri) [inline agent/agent.go:928]: Review: Unrelated fix(?). It seemed like a bad idea (to me) to perform a write on a network connection whilst holding the mutex, this shortens the span.

However, neither now, nor before, do we achieve perfect synchronization. In theory we should ensure that writes happening between writing `prevBuf` to conn and adding conn to `activeConns` are not lost. I did not try to fix this issue in this PR.

File: agent/agent.go [hunk 2]
```diff
@@ -946,16 +968,16 @@ func (a *agent) handleReconnectingPTY(ctx context.Context, msg codersdk.Reconnec
...
-			a.logger.Warn(ctx, "reconnecting pty buffer read error", slog.F("id", msg.ID), slog.Error(err))
-			return
+			logger.Warn(ctx, "read conn", slog.Error(err)
+			return nil
 		}
 		_, err = rpty.ptty.Input().Write([]byte(req.Data))
 		if err != nil {
-			a.logger.Warn(ctx, "write to reconnecting pty", slog.F("id", msg.ID), slog.Error(err))
-			return
+			logger.Warn(ctx, "write to pty", slog.Error(err)
```
Comment (mafredri) [inline agent/agent.go:979]: Review: It's unclear to my why some logs have been `Warn` whereas others are `Error`, I did however notice some tests fail unless I mimic the previous behavior exactly, which is what I've attempted here.

File: coderd/coderdtest/coderdtest.go [hunk 3]
```diff
@@ -604,6 +615,7 @@ func AwaitWorkspaceAgents(t *testing.T, client *codersdk.Client, workspaceID uui
 
 		return true
 	}, testutil.WaitLong, testutil.IntervalFast)
+	t.Logf("got workspace agents (workspace %s)", workspaceID)
```
Comment (mafredri) [inline coderd/coderdtest/coderdtest.go:618]: Review: These log messages help discern the order of events.

File: agent/agent.go [hunk 4]
```diff
@@ -798,38 +803,56 @@ func (a *agent) handleSSHSession(session ssh.Session) (retErr error) {
 	return cmd.Wait()
 }
 
-func (a *agent) handleReconnectingPTY(ctx context.Context, msg codersdk.ReconnectingPTYInit, conn net.Conn) {
+func (a *agent) handleReconnectingPTY(ctx context.Context, logger slog.Logger, msg codersdk.ReconnectingPTYInit, conn net.Conn) (retErr error) {
 	defer conn.Close()
 
 	connectionID := uuid.NewString()
+	logger = logger.With(slog.F("id", msg.ID), slog.F("connection_id", connectionID))
```
Comment (mafredri) [inline agent/agent.go:810]: Review: Addition of `connection_id` will help debug if logs are coming from one or multiple connections.

File: agent/agent.go [hunk 5]
```diff
@@ -300,10 +301,12 @@ func (a *agent) createTailnet(ctx context.Context, derpMap *tailcfg.DERPMap) (_
 		}
 	}()
 	if err = a.trackConnGoroutine(func() {
+		logger := a.logger.Named("reconnecting-pty")
```
Comment (mafredri) [inline agent/agent.go:304]: Review: Named logging seemed like it would be helpful here, even though it's not fully a separate service.

File: agent/agent.go [hunk 6]
```diff
@@ -101,8 +102,8 @@ func New(options Options) io.Closer {
 		filesystem:             options.Filesystem,
 		tempDir:                options.TempDir,
 	}
-	server.init(ctx)
-	return server
+	a.init(ctx)
+	return a
```
Comment (mafredri) [inline agent/agent.go:106]: Review: This just bothered me, lol.

Comment (kylecarbs) [reply agent/agent.go:106]: ;p I agree

File: agent/agent.go (hunk 1 above)
Comment (kylecarbs) [reply agent/agent.go:928]: Makes sense. It might be worth leaving that in a comment.

**Review by kylecarbs — APPROVED** [review]

File: agent/agent.go (hunk 2 above)
Comment (kylecarbs) [reply agent/agent.go]: That's unfortunate D: I believe they were _best effort_

**Review by kylecarbs — APPROVED** [review]
